Library/OcAppleKernelLib/PrelinkedContext.c

```diff
@@ -899,6 +909,129 @@ PrelinkedInjectComplete (
+      while (Symbols < SymbolsEnd) {
+        LengthOfPrelinkedSymbols += sizeof (LILU_PRELINKED_SYMBOLS_ENTRY) + Symbols->Length;
+        Symbols++;
+      }
+    }
+
+    Link = GetNextNode (&Context->PrelinkedKexts, Link);
+  }
+
+  // Allocate buffer and setup header
+  LiluPrelinkedSymbolsAddr = (UINTN)AllocateRuntimePool (LengthOfPrelinkedSymbols);
+  Buffer                   = (VOID *)LiluPrelinkedSymbolsAddr;
+  if (!Buffer) {
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  LiluHeader                  = (LILU_PRELINKED_SYMBOLS_HEADER *)Buffer;
+  LiluHeader->Version         = 0;
+  LiluHeader->Size            = LengthOfPrelinkedSymbols;
+  LiluHeader->NumberOfSymbols = NumSymbolsInPrelinked;
+
+  // Fill in the entries
+  CurEntry = (LILU_PRELINKED_SYMBOLS_ENTRY *)(LiluHeader + 1);
+  Link     = GetFirstNode (&Context->PrelinkedKexts);
+  Kext     = NULL;
+  while (!IsNull (&Context->PrelinkedKexts, Link)) {
+    Kext = GET_PRELINKED_KEXT_FROM_LINK (Link);
+    if (Kext->LinkedSymbolTable != NULL) {
+      NumSymbolsInKext = Kext->NumberOfSymbols;
+      Symbols          = Kext->LinkedSymbolTable;
+
+      SymbolsEnd = &Symbols[NumSymbolsInKext];
+      while (Symbols < SymbolsEnd) {
+        CurEntry->EntryLength      = sizeof (LILU_PRELINKED_SYMBOLS_ENTRY) + Symbols->Length;
+        CurEntry->SymbolValue      = Symbols->Value - 0xFFFFFF8000100000ULL; // KERNEL_HIB_VADDR
+        CurEntry->SymbolNameLength = Symbols->Length;
+        CopyMem (&CurEntry->SymbolName, Symbols->Name, Symbols->Length + 1);
+
+        CurEntry = (LILU_PRELINKED_SYMBOLS_ENTRY *)(((UINT8 *)CurEntry) + CurEntry->EntryLength);
+        Symbols++;
+      }
+    }
+
+    Link = GetNextNode (&Context->PrelinkedKexts, Link);
+  }
+
+  LiluPrelinkedSymbolsAddr64 = (UINT64)LiluPrelinkedSymbolsAddr;
+  DEBUG ((
+    DEBUG_INFO,
+    "OCAK: Lilu prelinked symbols are stored at 0x%llx with a size of 0x%x bytes and %d symbols\n",
+    LiluPrelinkedSymbolsAddr64,
+    LengthOfPrelinkedSymbols,
+    NumSymbolsInPrelinked
+    ));
+
+  // Expose the physical address of the LILU_PRELINKED_SYMBOLS struct via an EFI variable
+  Status = OcSetSystemVariable (
+             OC_LILU_PRELINKED_SYMBOLS_ADDR_VARIABLE_NAME,
+             OPEN_CORE_NVRAM_ATTR,
+             8,
+             (void *)&LiluPrelinkedSymbolsAddr64,
+             &Guid
+             );
+
+  if (EFI_ERROR (Status)) {
+    FreePool (Buffer);
+    return Status;
+  }
+
+  // Tell Lilu the amount of kexts to inject
+  Status = OcSetSystemVariable (
+             OC_LILU_KEXT_COUNT_VARIABLE_NAME,
+             OPEN_CORE_NVRAM_ATTR,
+             4,
+             (void *)&Context->LiluKextCount,
+             &Guid
+             );
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  return OcSetSystemVariable (
+           OC_LILU_BLOCK_INFO_ADDR_VARIABLE_NAME,
+           OPEN_CORE_NVRAM_ATTR,
+           8,
+           (void *)&Context->LiluBlockInfos,
+           &Guid
+           );
```

Comment:
All this may leave a broken, partial state where N vars succeeded to be set, but at least one didn't. Using one variable, as I suggested above, should resolve that though.